modules/ingest-common/src/main/java/org/opensearch/ingest/common/FingerprintProcessor.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.ingest.common;
+
+import org.opensearch.common.Nullable;
+import org.opensearch.common.hash.MessageDigests;
+import org.opensearch.ingest.AbstractProcessor;
+import org.opensearch.ingest.ConfigurationUtils;
+import org.opensearch.ingest.IngestDocument;
+import org.opensearch.ingest.Processor;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.opensearch.ingest.ConfigurationUtils.newConfigurationException;
+
+/**
+ * Processor that generating hash value for the specified fields or all fields in a document
+ */
+public class FingerprintProcessor extends AbstractProcessor {
+    public static final String TYPE = "fingerprint";
+    private static final Set<String> HASH_METHODS = Set.of("MD5", "SHA-1", "SHA-256", "SHA3-256");
+
+    // fields used to generate hash value
+    private final List<String> fields;
```

Comment:
I think my last concern is the `fields` & `includeAllFields`, usage of both properties is conflicting. As far as I can tell, we use different approach in other processors by relying in field patterns and exclusion lists (see `RemoveByPatternProcessor` fe):

```
    private final List<String> fieldPatterns;
    private final List<String> excludeFieldPatterns;
```

It is easier to reason about and configure, the `includeAllFields` would correspond to  `*` pattern, what do you think?

Reply:
Because field name can be a single star `*`:
```
{
  "test": {
    "mappings": {
      "properties": {
        "*": {
          "type": "text"
        }
      }
    }
  }
}
```
, so we may not use `*` to represent all fields, I think that's why we implement a new `remove_by_pattern` processor and introduce new parameters whose name contains `pattern`, but not make the parameter `field` in the `remove` processor support `*`, that may introduce confusion.

Reply:
I think we can add both inclusion and exclusion list, but seems other processors don't guarantee that empty list means all fields, for example, if the `field` or `exclude_field` in the `remove` processor is empty, the processor does nothing:
```
POST _ingest/pipeline/_simulate
{
  "pipeline": {
    "processors": [
      {
        "remove": {
          "field": []
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
        "uri": [
          "app","home","overview"
        ],
        "a":1
      }
    }
  ]
}
```
,response:
```
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "a": 1,
          "uri": [
            "app",
            "home",
            "overview"
          ]
        },
        "_ingest": {
          "timestamp": "2024-05-24T06:01:44.1073448Z"
        }
      }
    }
  ]
}
```
, so maybe we still need the flag `include_all_fields`.

Reply:
> I think we can add both inclusion and exclusion list, but seems other processors don't guarantee that empty list means all fields, 

We have a mix, I agree. Documenting would help for sure (the semantic of `RemoveProcessor` is somewhat different - it always need a field to be removed, I think empty field list should be rejected it looks like)

> , so maybe we still need the flag include_all_fields.

The flag would work, the reason I trying to get rid of it is to have non-conflicting and easy configuration. If use has no includes nor excludes (not provided or empty)  - all fields, otherwise - pick and choose.

 @msfroh do you have thoughts i

Reply:
:bow:  for doing the change, could you please also update description payloads accordingly, thank you

Reply:
Thank you, the pull request description would need some adjustments as well

Reply:
Thanks, I've done that.